### PR TITLE
Add pre-install checks

### PR DESCRIPTION
**Pull request by jan-zmeskal**
Check if default firewall manager is installed if user doesn't override this variable. More background info here: https://bugzilla.redhat.com/show_bug.cgi?id=1816975

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

File: tasks/pre_install_checks.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Gather facts on installed packages
+  package_facts:
+
```
Comment (mnecas) [inline tasks/pre_install_checks.yml]: Please add `manager: rpm`

File: defaults/main.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ ovirt_engine_setup_dwh_db_name: 'ovirt_engine_history'
 ovirt_engine_setup_dwh_db_user: 'ovirt_engine_history'
 ovirt_engine_setup_dwh_vacuum_full: false
 
-ovirt_engine_setup_firewall_manager: 'firewalld'
+ovirt_engine_setup_firewall_manager: '{{ default_firewall_manager }}'
```
Comment (mnecas) [inline defaults/main.yml]: What is the reason for this change and why use vars/main.yml? In this way, the user cannot edit the variable from vars.

File: tasks/pre_install_checks.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+- name: Fail when {{ default_firewall_manager }} is not installed
+  fail:
+    msg: '{{ ovirt_engine_setup_firewall_manager }} was chosen as default firewall manager but is not installed'
+  when: >
+    ovirt_engine_setup_firewall_manager == default_firewall_manager and
```
Comment (mnecas) [inline tasks/pre_install_checks.yml]: Remove this line.

File: tasks/pre_install_checks.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    default_firewall_manager not in ansible_facts.packages
```
Comment (mnecas) [inline tasks/pre_install_checks.yml]: Would need to check if the value is `null` to skip it because of this use-case in README `empty value to skip firewall configuration`

File: tasks/pre_install_checks.yml (hunk 1 above)
Comment (mnecas) [reply tasks/pre_install_checks.yml]: and also `no_log: yes`

File: defaults/main.yml (hunk 2 above)
Comment (jan-zmeskal) [reply defaults/main.yml]: This is exactly the purpose. We want user to override variable `ovirt_engine_setup_firewall_manager` which is indeed firewall manager of their choice. That is why `ovirt_engine_setup_firewall_manager` is in default/main.yml - because we **expect** users to override the default values.

However, no matter what will the user provide to `ovirt_engine_setup_firewall_manager`, we always want to treat `firewalld` as the default manager. This is something that user doesn't choose. That's why I put `default_firewall_manager` into vars/mail.yml.

Alternate approach would be to just keep it in default/main.yml. But then, in all the conditional comparisons, I would need to spell out "firewalld" explicitly. For example `when: ovirt_engine_setup_firewall_manager == firewalld`. This seems like a bad design from maintainability point of view.

File: tasks/pre_install_checks.yml (hunk 1 above)
Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: I can add `manager: rpm` but then I might need to add additional conditional that checks that we are on RHEL-based system, right? Or is that not necessary?

File: tasks/pre_install_checks.yml (hunk 3 above)
Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: Why remove? This is the whole point of this PR. If the use chose different firewall manager, then if don't care whether the default one is installed or not

File: tasks/pre_install_checks.yml (hunk 4 above)
Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: I'm afraid I don't understand this comment, could you elaborate?

**Comment by jan-zmeskal** [comment]
> Not sure if this logic should be done in this role. IMHO shoud be done in engine-setup.
> If it will end up here we could add another parameter which would allow automatic install of the firewall instead of fail.

I guess that would be ideal. But Michal told me to [post a patch](https://bugzilla.redhat.com/show_bug.cgi?id=1816975#c1). And since I'm not able to add this functionality to `engine-setup`, I made the PR here.

File: tasks/pre_install_checks.yml (hunk 1 above)
Comment (mnecas) [reply tasks/pre_install_checks.yml]: It is unnecessary.

File: defaults/main.yml (hunk 2 above)
Comment (mnecas) [reply defaults/main.yml]: So if I understand it correctly the plan is to drop all options for `ovirt_engine_setup_firewall_manager` and always use `firewalld`? Because right now users can choose between `firewalld`, `iptables` or `null`.

Comment (mnecas) [reply defaults/main.yml]: You could just skip the `ovirt_engine_setup_firewall_manager == firewalld` so it would check only if the package is in the facts. 
example:
```yml
      when: 
        - ovirt_engine_setup_firewall_manager not in ansible_facts.packages
        - ovirt_engine_setup_firewall_manager
```
this works when the `ovirt_engine_setup_firewall_manager` is  `null`, `firewalld` and `iptables`

You don't need to specify extra variables.

File: tasks/pre_install_checks.yml (hunk 1 above)
Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: I'll add it I just wonder: why the `no_log` statement here? Is this considered sensitive?

**Comment by jan-zmeskal** [comment]
@mnecas I posted a new commit. I was finally able to understand what did you mean to say to me :-D I think now it should be much more compliant with your ideas

File: tasks/pre_install_checks.yml (hunk 1 above)
Comment (mnecas) [reply tasks/pre_install_checks.yml]: Both reasons.

File: tasks/pre_install_checks.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  package_facts:
+    manager: rpm
+  no_log: true
+
+- name: Fail when firewall manager is not installed
+  fail:
+    msg: '{{ ovirt_engine_setup_firewall_manager }} was chosen as a firewall manager but is not installed'
+  when:
+    - ovirt_engine_setup_firewall_manager not in ansible_facts.packages
+    - ovirt_engine_setup_firewall_manager|bool
```
Comment (mnecas) [inline tasks/pre_install_checks.yml]: remove the `|bool` (it converts `"True"` to `True`)

Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: Removing `bool` triggers the following warning when `ovirt_engine_setup_firewall_manager` is `null`:
```
[DEPRECATION WARNING]: evaluating None as a bare variable, this behaviour will go away and you might need to add |bool to the expression in the future. Also see CONDITIONAL_BARE_VARS 
configuration toggle.. This feature will be removed in version 2.12. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```
Are wa okay with that?

Comment (mnecas) [reply tasks/pre_install_checks.yml]: Oh did not know about this use case with null... better to keep the bool

Comment (mnecas) [reply tasks/pre_install_checks.yml]: done some testing and think it would break the logic because if there is some other value then "True" it's converted to False, so when you would put for example firewalld it would not pass

try to use `is not none`

Comment (jan-zmeskal) [reply tasks/pre_install_checks.yml]: The check for `is not none` seems to be working nice. I ran a series of tests, see here: https://pastebin.com/UmmEYG2x

During the tests, I disabled most of the role's execution and kept only tasks/pre_install_checks.yml. All of the different cases now behave as we expect. The only one that's a little quirky is the last one - empty string. It fails because "" of course is not among installed packages. However, I think that's fine because we instruct users to provide empty value if they want to skip the firewall configuration.
